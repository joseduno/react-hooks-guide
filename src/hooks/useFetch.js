import { useEffect, useRef, useState } from "react"


const useFetch = (url) => {

    // El valor de isMounted se almacena en isMounted.current
    const isMounted = useRef(true)
    const [state, setState] = useState({data: null, loading: true, error: null})

    useEffect(() => {
        return () => {
            isMounted.current = false
        }
    }, [])

    useEffect(() => {

        setState(s => ({...s, loading:true}))
        
        fetch(url)
            .then(resp => resp.json())
            .then(data =>{
                if (isMounted.current) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    })
                } else {
                    console.log('setState no se llamo')
                }
            })
            .catch(e => {
                setState({data: null, loading: false, error: 'No se pudo cargar la info'})
            })
        
    }, [url])

    return state

}


export default useFetch