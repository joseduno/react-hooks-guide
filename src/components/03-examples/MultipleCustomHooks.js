import React from 'react'
import useCounter from '../../hooks/useCounter'
import useFetch from '../../hooks/useFetch'

const MultipleCustomHooks = () => {

    const {state, increment} = useCounter(1)

    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${state}`)

    // data incialmente es null, !null returna true !!null retorna false
    // && if corto sin sentencia else
    const {author, quote} = !!data && data[0]

    return (
        <div className='container mt-5'>
            <h1>MultipleCustomHooks</h1>
            <h2>Breaking Bad Quotes</h2>
            <p>
                Utilizando e interactuando 2 custom hooks, useCounter y <b>useFetch</b>. 
                Ã‰ste ultimo hook muestra el uso del useEffect, peticiones a una API de
                de forma asincrona y otro ejemplo de uso del return en un useEffect
            </p>

            {loading ? (
                <div className='alert alert-info text-center'>Loading...</div>
                ) : (
                <blockquote className='blockquote'>
                    <p>{quote}</p>
                    <footer className='blockquote-footer'>{author}</footer>
                </blockquote>
            )}

            <button className='btn btn-primary' onClick={()=>increment(1)}>
                Next quote
            </button>
            
            <hr />
        </div>
    )
}

export default MultipleCustomHooks
