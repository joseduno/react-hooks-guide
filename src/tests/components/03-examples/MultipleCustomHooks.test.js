import { shallow } from "enzyme";
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import useFetch from "../../../hooks/useFetch";
import useCounter from "../../../hooks/useCounter";
jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')


/*
jest.mock('../../hooks/useFetch') junto con useFetch.mockReturnValue({data: null, loading: true, error: null})
se utiliza para falsear la implementacion del hook, la intencion es probar el componente 
de forma aislada a los hooks que utiliza. Deben haber pruebas especificas para los hooks
tambien aislada de los componentes que lo usan
*/


describe('Pruebas en <MultipleCustomHooks />', () => {

    
    test('debe mostrarse correctamente sin data', () => {
        useCounter.mockReturnValue({
            state: 1,
            increment: jest.fn()
        })
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })
        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper).toMatchSnapshot()
    })

    test('debe mostrarse correctamente con data', () => {
        useCounter.mockReturnValue({
            state: 1,
            increment: jest.fn()
        })
        useFetch.mockReturnValue({
            data: [{author: 'jose', quote: 'hola'}],
            loading: false,
            error: null
        })
        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper.find('.alert').exists()).toBe(false)
        expect(wrapper.find('p').at(1).text().trim()).toBe('hola')
        expect(wrapper.find('footer').text().trim()).toBe('jose')
        
    })
    
})
