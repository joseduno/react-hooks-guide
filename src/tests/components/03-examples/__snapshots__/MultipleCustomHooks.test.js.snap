// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe mostrarse correctamente sin data 1`] = `
<div
  className="container mt-5"
>
  <h1>
    MultipleCustomHooks
  </h1>
  <h2>
    Breaking Bad Quotes
  </h2>
  <p>
    Utilizando e interactuando 2 custom hooks, useCounter y 
    <b>
      useFetch
    </b>
    . Ã‰ste ultimo hook muestra el uso del useEffect, peticiones a una API de de forma asincrona y otro ejemplo de uso del return en un useEffect
  </p>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
  <hr />
</div>
`;
